test/testtokenize.cpp

```diff
@@ -7904,6 +7907,23 @@ class TestTokenizer : public TestFixture {
         ASSERT_EQUALS("b(", testAst("class a { void b() & {} };"));
     }
 
+    //Verify that returning a newly constructed object generates the correct AST even when the class name is scoped
+    //Addresses https://trac.cppcheck.net/ticket/9700
+    void astnewscoped() {
+        ASSERT_EQUALS("(return (new A))", testAst("return new A;", false, true));
+        ASSERT_EQUALS("(return (new (( A)))", testAst("return new A();", false, true));
+        ASSERT_EQUALS("(return (new (( A true)))", testAst("return new A(true);", false, true));
+        ASSERT_EQUALS("(return (new (:: A B)))", testAst("return new A::B;", false, true));
+        ASSERT_EQUALS("(return (new (( (:: A B))))", testAst("return new A::B();", false, true));
+        ASSERT_EQUALS("(return (new (( (:: A B) true)))", testAst("return new A::B(true);", false, true));
+        ASSERT_EQUALS("(return (new (:: (:: A B) C)))", testAst("return new A::B::C;", false, true));
+        ASSERT_EQUALS("(return (new (( (:: (:: A B) C))))", testAst("return new A::B::C();", false, true));
+        ASSERT_EQUALS("(return (new (( (:: (:: A B) C) true)))", testAst("return new A::B::C(true);", false, true));
+        ASSERT_EQUALS("(return (new (:: (:: (:: A B) C) D)))", testAst("return new A::B::C::D;", false, true));
+        ASSERT_EQUALS("(return (new (( (:: (:: (:: A B) C) D))))", testAst("return new A::B::C::D();", false, true));
+        ASSERT_EQUALS("(return (new (( (:: (:: (:: A B) C) D) true)))", testAst("return new A::B::C::D(true);", false, true));
```

Comment:
@danmar I have to admit your z3 suggestion does seem pretty readable and I can't come up with a concrete case where the ```(``` is ambiguous.